### DISCUSSION TITLE
model module 분리 관련 논의

### DISCUSSION THREAD

**Discussion by seonpilKim** [zoopi-palette/zoopi-spring-backend#30] [Discussion]
[comment by #25](https://github.com/zoopi-palette/zoopi-spring-backend/pull/25#discussion_r933968955)에서 시작된 주제입니다.
사실 domain 모듈에 Response 클래스가 이거 말고도 또 있는데요, ErrorResponse입니다.
또 domain 모듈에 OAtuh2SuccessHandler 클래스를 보시면, api 모듈에 있는 ResultResponse 클래스를 참조하지 못해서 동일한 형태로 Response 클래스를 만들어서 사용하기도 하는데, 이런 부분들이 구현하면서 좀 거슬렸지만 모른척 했던(?) 부분들인데, 마침 찬준님께서 [comment](https://github.com/zoopi-palette/zoopi-spring-backend/pull/25#discussion_r933968955)를 남겨주셨으니, 확실히 짚고 넘어가는게 좋을 것 같아서 고민좀 해봤는데요,

혹시 `module-model`을 따로 분리하고, domain 모듈과 api 모듈이 model 모듈을 의존하는 형태는 어떻게 생각하시나요?
해당 모듈에서 api interface랑 request/response class를 관리하고, api 모듈에서 api interface를 상속받아 controller를 구현하고, request/response class는 필요한 모듈(api, domain)에서 자유롭게 사용하는 방식입니다.
글로만 써서 제 의도가 잘 전달될진 모르겠네요😅

아래는 간단한 예시입니다.
### module-model의 MemberAuthApi interface
```java
@Validated
@RequestMapping("/auth")
@Api(tags = "회원 인증 API")
public interface MemberAuthApi {

	@ApiOperation(value = "아이디 유효성 검사")
	@ApiResponses({
		@ApiResponse(code = 200, message = "[R-M001] 사용 가능한 아이디입니다.", response = MemberAuthDto.ValidationResponse.class)
	})
	@ApiImplicitParam(name = "username", value = "아이디", required = true, example = "zoopi")
	@PostMapping("/username/validate")
	ResponseEntity<ResultResponse> validateUsername(@RequestParam @Size(max = 30) String username);

}
```
### module-api의 MemberAuthController
```java
@RestController
@RequiredArgsConstructor
public class MemberAuthController implements MemberAuthApi {

	private final MemberService memberService;

	@Override
	public ResponseEntity<ResultResponse> validateUsername(String username) {
		final boolean isAvailable = memberService.isAvailableUsername(username);
		final ResultResponse response = ValidationResponse.fromValidatingUsername(username, isAvailable);
		return ResponseEntity.ok(response);
	}
}
```

이렇게 하면 장점은, 구현하기 전 api 명세서를 먼저 빠르게 설계해서 프론트/클라이언트 분들께 전달할 수 있고, controller단에서 api 명세서에 관련된 지저분한 코드를 더 이상 보지 않아도 됩니다. (api 모듈에서 swagger 의존성을 떼어낼 수 있습니다.)
또한, 이전의 문제였던 response 클래스를 domain 모듈에서 관리할 필요 없이 호출이 가능하게 됩니다.

model 모듈 구조는 아래와 같습니다.
```
└──🔹zoopi
      ├──🔹module-model
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂swagger
      │     │     │     │     │     └── 🔸SwaggerConfig.java
      │     │     │     │     ├── 📂model
      │     │     │     │     │     │     ├── 📂request 
      │     │     │     │     │     │     │     └── 🔸SignupRequest.java
      │     │     │     │     │     │     ├── 📂response
      │     │     │     │     │     │     │     └── 🔸ValidationResponse.java
      │     │     │     │     │     ├── 🔸ErrorCode.enum
      │     │     │     │     │     ├── 🔸ErrorResponse.java
      │     │     │     │     │     ├── 🔸ResultCode.enum
      │     │     │     │     │     └── 🔸ResultResponse.java
      │     │     │     │     ├── 📂api
      │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     ├── 🔸MemberApi.java
      │     │     │     │     │     │     └── 🔸AuthApi.java
```

**Comment by slolee** [comment]
음..... 이러한 의존관계를 두는게 제 2의 common 모듈을 만드는 것 같아서 조금 복잡도가 올라갈까 걱정이 되네요!
지금 구조에서 질문드리고 싶은게 있는데요. 

혹시 지금 구조가 module-api 에 컨트롤러(Presentation Layer) 만 존재하고 Service Layer 가 module-domain 모듈로 빠져있는데 이러게 설계하신 의도가 있으실까요? 제가 잘못 생각하고 있는것일수도 있지만 Service Layer 가 module-api 로 올라오는게 조금 더 자연스럽지 않을까..? 하는 생각이 조금 있습니다!

Service Layer 는 요청에 따른 비즈니스 로직 처리를 담당하는데 "요청"에 따라서 이기 때문에 Domain 보다는 Controller 에 의존적으로 존재하는게 맞을것 같다는 생각이 듭니다! 혹시 이 부분에 대해서 어떻게 생각하실까요?

**Reply by seonpilKim** [reply]
제가 생각하는 멀티 모듈의 목적은 역할에 따라 모듈을 분리하는 것이라 생각합니다.
먼저 불필요한 의존성을 줄여주고, 모듈을 분리함으로써 각 모듈의 역할과 책임이 분명해지기에 개발 및 유지보수에 이점이 있다고 생각합니다.
- `common 모듈`: 자바를 제외하면 어떠한 의존성도 없으며, 모든 모듈에서 공통적으로 사용되는 파일들을 관리하는 모듈
    > ex) Utils, Constants
- `model 모듈`: 클라이언트의 요청/응답 데이터를 정의하고, api 명세서(interface)를 관리하는 모듈
- `api 모듈`: 클라이언트의 요청과 응답을 처리하는 모듈
- `domain 모듈`: 시스템의 도메인을 다루는 모듈
    - 현재는 repository가 service와 동일한 모듈에 위치하지만, rdb 외에 nosql도 도입하는 경우 모듈을 더 잘게 쪼개려고 생각하고 있었습니다.
    > ex) module-domain에서 module-rds, module-redis 분리<br> 
![image](https://user-images.githubusercontent.com/68049320/183235540-b80c14f2-a855-49d6-8af5-5227288112aa.png)

**Reply by slolee** [reply]
네네 말씀해주신 내용은 모두 동의하고 저번에 공유해주셔서 저도 인상깊게 읽은 내용입니다 👍👍 

다만 지금 보내주신 그림과 현재 주피 프로젝트가 구성되어있는 모듈에는 조금 차이가 있는것 같아요.

위 그림대로라면 도메인 모듈에는 도메인에 대한 정의와 그와 관련된 서비스만 존재해야 합니다. 그리고 그 도메인들을 가져다 사용하는건 서비스 계층이 되고, 그 서비스 계층이 현재 도메인 모듈에 함께 존재하는 상태인것 같아요. 

이렇게되면 해당 도메인을 다른 시스템(배치와같은)에서 활용하고자 할 때 현재 도메인 모듈에 함께 정의되어있는 서비스를 사용해야하는데 API 에서 하고자하는 비즈니스 로직과 달라질 가능성이 매우 크고 그러면 또 그에 대한 서비스를 별도로 개발해야하는 상황이생길거 같습니다.

그래서 도메인 모델을 가져와서 사용하고 이를 이용해 비즈니스 로직을 구현하는 부분은 module-api 에 와야하는게 아닌가 하는 생각이었습니다.

도메인 모듈을 별도로 분리하는 이유가 api, batch, admin 등 이런 다양한 서비스가 공통된 도메인 모델과 도메인 기능을 활용하기 위함이지, 공통된 비즈니스 로직을 사용하기 위함이 아니라고 생각합니다.

> 모듈을 분리함으로써 각 모듈의 역할과 책임이 분명해지기에 개발 및 유지보수에 이점이 있다고 생각합니다.

제 생각을 정리하자면,,, 말씀해주신 부분에서 모듈의 역할과 책임을 분명히하기 위해서 비즈니스 로직을 윗 계층으로 올려야하는게 아닐까..? 입니다!

---

이건 TMI 이지만,,, DDD 와 JPA 를 동시에 활용하게되는 경우 제대로 의존성관리가 안되는 문제가 있습니다.
DDD 관점에서 도메인 모델은 Infrastructure(이하 Infra) Layer 를 의존해서는 안되지만, JPA 를 사용하게되면 도메인 모델에 다양한 JPA 관련 어노테이션이 추가되고 다시말하면 Infra Layer 를 의존하게 되는거죠. 그래서 이상적인 DDD 를 구현하려면 JPA 에서 사용하는 Entity 와 도메인 모델을 위한 Entity 를 별도로 구성해야 하지만 오버 엔지니어링이 될 확률이 매우 크기 때문에 트레이드 오프를 하는 경우가 많습니다. 현재 주피 프로젝트도 그런 상태라고 이해하고 있습니다.

그래서 말씀하신대로 module-domain 에서 module-rds 를 분리해내는게 좋은 방향인가..? 에 대한 의문점은 있습니다. 이상적으로는 맞는 방향이지만 설계적인 부분에 매몰되어 너무 오버엔지니어링을 하는건 아닐까..? 하는 고민도 있습니다.

**Reply by seonpilKim** [reply]
찬준님 코멘트를 읽으면서 많은 생각을 하게 되었습니다. 감사합니다😊
고민하는 시간이 길어지다보니 답변이 늦어진 점은 정말 죄송합니다ㅠ_ㅠ

일단 api, batch, admin에 대한 각 비즈니스 로직이 다를테니, 이를 분리하는 건 좋은 것 같습니다👍
다만 api 모듈에 controller, service 계층을 함께 두는 것 보다 service 모듈을 따로 분리하는 건 어떨까요?
아무래도 controller는 앞단과 통신하는 역할을 수행하고, service는 비즈니스 로직을 담당하니까 둘을 분리하고 싶은 생각이 계속 들어서요 ㅠ_ㅠ
그리고 service 계층에서만 의존하면 되는 것들이 controller 계층에서도 불필요하게 함께 의존해야 하는 상황도 생길 것 같습니다.

> 이건 TMI 이지만,,, DDD 와 JPA 를 동시에 활용하게되는 경우 제대로 의존성관리가 안되는 문제가 있습니다.
DDD 관점에서 도메인 모델은 Infrastructure(이하 Infra) Layer 를 의존해서는 안되지만, JPA 를 사용하게되면 도메인 모델에 다양한 JPA 관련 어노테이션이 추가되고 다시말하면 Infra Layer 를 의존하게 되는거죠. 그래서 이상적인 DDD 를 구현하려면 JPA 에서 사용하는 Entity 와 도메인 모델을 위한 Entity 를 별도로 구성해야 하지만 오버 엔지니어링이 될 확률이 매우 크기 때문에 트레이드 오프를 하는 경우가 많습니다. 현재 주피 프로젝트도 그런 상태라고 이해하고 있습니다.

음 확실히 찬준님 말씀대로 별도로 구성을 하게 되면 DB 구현체를 변경할 때 doamin 영역을 유지할 수 있다는 장점은 있지만 저희는 JPA를 고정으로 사용하는 것을 전제로 깔고 간다면, 굳이 infra를 별도로 나눌 필요는 없지 않을까라는 생각이 드네요..!
만약 데이터 접근 기술을 정말 분리해야 한다면, 그 때 한 번 리팩토링하는게 좋을 것 같기도 합니다!

> 그래서 말씀하신대로 module-domain 에서 module-rds 를 분리해내는게 좋은 방향인가..? 에 대한 의문점은 있습니다. 이상적으로는 맞는 방향이지만 설계적인 부분에 매몰되어 너무 오버엔지니어링을 하는건 아닐까..? 하는 고민도 있습니다.

아아.. 사실 rds, redis 분리 말고도 infra(aws, sms)도 분리하려고 생각하고 있었는데.. 너무 오버일까요..? 각 모듈별로 필요한 의존성만 가져오게 하려면 어쩔 수 없이 나눠야 할 것 같은데.. 이 부분은 방향성 잡기가 너무 어렵더라구요 ㅠㅠ

일단 제가 생각하고 있는 구조도 보여드리면 좋을 것 같아서 아래에 추가해봅니다..!

#### 예상 모듈 의존관계
![image](https://user-images.githubusercontent.com/68049320/183245886-a5b59e66-e26a-4951-bf6f-e3a89464d3c6.png)

#### 개인적인 예상 전체 구조
```
└──🔹zoopi
      ├──🔹module-model
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂swagger
      │     │     │     │     │     └── 🔸SwaggerConfig.java
      │     │     │     │     ├── 📂api
      │     │     │     │     │     ├── 📂admin
      │     │     │     │     │     ├── 📂client
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     └── 🔸MemberAuthApi.java
      │     │     │     │     ├── 📂model
      │     │     │     │     │     ├── 📂admin
      │     │     │     │     │     ├── 📂client
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 📂request
      │     │     │     │     │     │     │     │     └── 🔸MemberAuthRequest.java
      │     │     │     │     │     │     │     ├── 📂response
      │     │     │     │     │     │     │     │     └── 🔸MemberAuthResponse.java
      │     │     │     │     │     ├── 🔸ErrorCode.enum
      │     │     │     │     │     ├── 🔸ErrorResponse.java
      │     │     │     │     │     ├── 🔸ResultCode.enum
      │     │     │     │     │     └── 🔸ResultResponse.java

      ├──🔹module-application
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     │     ├── 📂advice
      │     │     │     │     │     │     └── 🔸GlobalExceptionHandler.java
      │     │     │     │     │     ├── 📂admin
      │     │     │     │     │     ├── 📂batch
      │     │     │     │     │     ├── 📂client
      │     │     │     │     │     │     ├── 📂controller
      │     │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     │     └── 🔸MemberAuthController.java
      │     │     │     │     └── 🔸ZoopiApplictaion.java

      ├──🔹module-service
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂admin
      │     │     │     │     ├── 📂batch
      │     │     │     │     ├── 📂client
      │     │     │     │     │     ├── 📂service
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 🔸MemberService.java
      │     │     │     │     │     │     │     ├── 🔸SnsAccountService.java
      │     │     │     │     │     │     │     ├── 📂dto
      │     │     │     │     │     │     │     └── 📂exception
      │     │     │     │     ├── 📂exception
      │     │     │     │     │     ├── 🔸BusinessException.java
      │     │     │     │     │     ├── 🔸EntityNotFoundException.java
      │     │     │     │     │     ├── 🔸...
      │     │     │     │     │     └── 🔸InvalidArgumentException.java
      │     │     │     │     ├── 📂security
      │     │     │     │     │     ├── 📂oauth2
      │     │     │     │     │     ├── 📂handler
      │     │     │     │     │     ├── 📂jwt
      │     │     │     │     │     ├── 📂matcher
      │     │     │     │     │     ├── 📂exception
      │     │     │     │     │     ├── 🔸WebSecurityConfig.java
      │     │     │     │     │     ├── 🔸PasswordEncoderConfig.java

      ├──🔹module-domain
      │     ├──🔹module-domain-redis
      │     ├──🔹module-domain-rds
      │     │     └── 📂src
      │     │     │     └── 📂main
      │     │     │     │     └── 📂java/com/zoopi
      │     │     │     │     │     ├── 📂domain
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 📂entity
      │     │     │     │     │     │     │     ├── 📂vo 
      │     │     │     │     │     │     │     ├── 📂dto 
      │     │     │     │     │     │     │     └── 📂repository

      ├──🔹module-infra
      │     ├──🔹module-infra-aws
      │     ├──🔹module-infra-sms
      │     │     └── 📂src
      │     │     │     └── 📂main
      │     │     │     │     └── 📂java/com/zoopi
      │     │     │     │     │     ├── 📂sms
      │     │     │     │     │     │     ├── 🔸NaverSensClient.java
      │     │     │     │     │     │     └── 🔸SmsClient.java

      ├──🔹module-common
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂util
      │     │     │     │     │     ├── 🔸Constants.java
      │     │     │     │     │     ├── 🔸...
      │     │     │     │     │     └── 🔸AuthenticationCodeUtils.java
```

**Comment by slolee** [comment]
회사일이 매우 바빠서 확인이 늦었습니다! 죄송합니다 ㅠㅠ

말씀하신대로 Presentation Layer 는 외부로부터의 요청을 받는 역할을 하고, Service Layer 는 비즈니스 로직을 처리하는 역할을 하도록 모듈을 나눠도 상관은 없습니다만... 프로젝트의 규모가 크지않고 모듈 복잡도가 올라갈것같다는 생각이 조금은 남아있는것 같아요 ㅠㅠ

해당 비즈니스 로직은 module-api 를 통해 들어온 요청에 대한 비즈니스 로직이고, 다른 곳에서 재사용이 불가능한 형태가 될 가능성이 높아서 저한테는 나눴을 때 얻을 수 있는 이득보다 복잡도가 올라감으로써 얻는 손해가 좀 더 크다고 느껴지는것 같아요.

이 부분은 선택적인 부분이라 선필님이 의도하신 방향으로 진행해주셔도 됩니다 :) 

> 아아.. 사실 rds, redis 분리 말고도 infra(aws, sms)도 분리하려고 생각하고 있었는데.. 너무 오버일까요..? 각 모듈별로 필요한 의존성만 가져오게 하려면 어쩔 수 없이 나눠야 할 것 같은데.. 이 부분은 방향성 잡기가 너무 어렵더라구요 ㅠㅠ

sms 와 같이 정확히 외부만 바라보는 경우 별도의 client 모듈(Infra Layer)로 나눠도 괜찮다고 생각해요. 이를 나눔으로써 해당 부분의 코드의 의존성을 좀더 깐깐하게 확인할 수 있을것 같아요. 물론 지금처럼 별도의 폴더로만 나눠서 진행해도 크게 무리는 없다고 생각합니다ㅎㅎ 현재처럼 폴더로 나눠놓고 진행하다가 나중에 언제든지 필요에의해서 별도 모듈로 빼낼수 있으니까요! 대신 지금처럼하면 의존성을 좀더 꼼꼼하게 확인하긴 해야할것 같아요. 도메인 모듈에 같이있다보니 도메인에 의존하기 쉬운상태인것 같아서요 ㅎㅎ

**Reply by seonpilKim** [reply]
> 프로젝트의 규모가 크지않고 모듈 복잡도가 올라갈것같다는 생각이 조금은 남아있는것 같아요 ㅠㅠ
해당 비즈니스 로직은 module-api 를 통해 들어온 요청에 대한 비즈니스 로직이고, 다른 곳에서 재사용이 불가능한 형태가 될 가능성이 높아서 저한테는 나눴을 때 얻을 수 있는 이득보다 복잡도가 올라감으로써 얻는 손해가 좀 더 크다고 느껴지는것 같아요.

(재사용성 낮음, 모듈 복잡성 증가 vs 의존성 및 책임 분리) 을 두고 고민을 많이 해보았는데, 말씀해 주신 것처럼 이득보다 손해가 크다는 느낌이 드는 것 같더라구요
그래서 service 모듈을 application 모듈에 합치고, 패키지를 따로 분리하여 구분하는 것으로 하는 방향으로 구조를 바꿔보았습니다!

괜찮으시다면 모듈 구조 변경 이슈를 추가하고 작업하도록 하겠습니다.

```
└──🔹zoopi
      ├──🔹module-model
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂swagger
      │     │     │     │     │     └── 🔸SwaggerConfig.java
      │     │     │     │     ├── 📂api
      │     │     │     │     │     ├── 📂admin
      │     │     │     │     │     ├── 📂client
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     └── 🔸MemberAuthApi.java
      │     │     │     │     ├── 📂model
      │     │     │     │     │     ├── 📂admin
      │     │     │     │     │     ├── 📂client
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 📂request
      │     │     │     │     │     │     │     │     └── 🔸MemberAuthRequest.java
      │     │     │     │     │     │     │     ├── 📂response
      │     │     │     │     │     │     │     │     └── 🔸MemberAuthResponse.java
      │     │     │     │     │     ├── 🔸ErrorCode.enum
      │     │     │     │     │     ├── 🔸ErrorResponse.java
      │     │     │     │     │     ├── 🔸ResultCode.enum
      │     │     │     │     │     └── 🔸ResultResponse.java

      ├──🔹module-application
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂advice
      │     │     │     │     │     └── 🔸GlobalExceptionHandler.java
      │     │     │     │     ├── 📂admin
      │     │     │     │     ├── 📂batch
      │     │     │     │     ├── 📂client
      │     │     │     │     │     ├── 📂controller
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     └── 🔸MemberAuthController.java
      │     │     │     │     │     ├── 📂service
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 🔸MemberService.java
      │     │     │     │     │     │     │     ├── 🔸SnsAccountService.java
      │     │     │     │     │     │     │     ├── 📂dto
      │     │     │     │     │     │     │     └── 📂exception
      │     │     │     │     ├── 📂exception
      │     │     │     │     │     ├── 🔸BusinessException.java
      │     │     │     │     │     ├── 🔸EntityNotFoundException.java
      │     │     │     │     │     ├── 🔸...
      │     │     │     │     │     └── 🔸InvalidArgumentException.java
      │     │     │     │     ├── 📂security
      │     │     │     │     │     ├── 📂oauth2
      │     │     │     │     │     ├── 📂handler
      │     │     │     │     │     ├── 📂jwt
      │     │     │     │     │     ├── 📂matcher
      │     │     │     │     │     ├── 📂exception
      │     │     │     │     │     ├── 🔸WebSecurityConfig.java
      │     │     │     │     │     ├── 🔸PasswordEncoderConfig.java
      │     │     │     │     └── 🔸ZoopiApplictaion.java

      ├──🔹module-domain
      │     ├──🔹module-domain-redis
      │     ├──🔹module-domain-rds
      │     │     └── 📂src
      │     │     │     └── 📂main
      │     │     │     │     └── 📂java/com/zoopi
      │     │     │     │     │     ├── 📂domain
      │     │     │     │     │     │     ├── 📂member
      │     │     │     │     │     │     │     ├── 📂entity
      │     │     │     │     │     │     │     ├── 📂vo 
      │     │     │     │     │     │     │     ├── 📂dto 
      │     │     │     │     │     │     │     └── 📂repository

      ├──🔹module-infra
      │     ├──🔹module-infra-aws
      │     ├──🔹module-infra-sms
      │     │     └── 📂src
      │     │     │     └── 📂main
      │     │     │     │     └── 📂java/com/zoopi
      │     │     │     │     │     ├── 📂sms
      │     │     │     │     │     │     ├── 🔸NaverSensClient.java
      │     │     │     │     │     │     └── 🔸SmsClient.java

      ├──🔹module-common
      │     └── 📂src
      │     │     └── 📂main
      │     │     │     └── 📂java/com/zoopi
      │     │     │     │     ├── 📂util
      │     │     │     │     │     ├── 🔸Constants.java
      │     │     │     │     │     ├── 🔸...
      │     │     │     │     │     └── 🔸AuthenticationCodeUtils.java
```

![image](https://user-images.githubusercontent.com/68049320/184476392-db38733f-e423-4458-885e-4165a60aae42.png)

**Reply by slolee** [reply]
넵 감사합니다 :) 👍 👍 👍
